specs/phase1/beacon-chain.md

```diff
@@ -886,9 +886,11 @@ def process_crosslink_for_shard(state: BeaconState,
         for attestation in transition_attestations:
             participants = get_attesting_indices(state, attestation.data, attestation.aggregation_bits)
             transition_participants = transition_participants.union(participants)
-            assert attestation.data.shard_head_root == shard_transition.shard_data_roots[
-                len(shard_transition.shard_data_roots) - 1
-            ]
+            if len(shard_transition.shard_states) > 0:
+                # Is the `shard_transition` candidate
+                last_offset_index = len(shard_transition.shard_states) - 1
+                shard_head_root = shard_transition.shard_states[last_offset_index].latest_block_root
+                assert attestation.data.shard_head_root == shard_head_root
```

Comment:
This assertion doesn't seem correct (nor was it prior to the recent change).

The single `shard_transition` is passed in to this function (the winning one included in the block) does not have to be reflected in all of the attestations included in the block. This assertion is forcing all attestations in the block to agree with the shard block head of the winning shard transition (if there is one).

I think what we want here is to make sure that `shard_head_block` votes and `shard_transition` votes are coherent. We can only do this for attestations that voted for the winning shard transition

Reply:
As for the `if len(shard_transition.shard_states) > 0:` part of this... I wish there was a cleaner way. Let's get the above fixed then we can discuss it again

Reply:
You're right. To make it correct, it can be checked after `enough_online_stake is True`.

```python
        # If not enough stake, try next transition root
        if not enough_online_stake:
            continue

        # Attestation <-> shard transition consistency
        assert shard_transition_root == hash_tree_root(shard_transition)

        # Check `shard_head_root` of the winning attestations
        last_offset_index = len(shard_transition.shard_states) - 1
        shard_head_root = shard_transition.shard_states[last_offset_index].latest_block_root
        for attestation in transition_attestations:
            assert attestation.data.shard_head_root == shard_head_root
```

@djrtwo @terencechain mind taking another look?

Reply:
`assert shard_transition != ShardTransition()` this could cause the chain to stall if a corrupt committee voted on `ShardTransition()` unless the node was looking for this type of attack and not including attestations because of what might happen. 

We want to add add in `validate_attestation` under `if is_on_time_attestation`: `assert attestation.data.shard_transition_root != hash_tree_root(ShardTransition())`.

We also kind of handle this in [verify_empty_shard_transition](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase1/beacon-chain.md#verify_empty_shard_transition) already

Not sure. Need to think about these corner cases a bit more. We definitely want the chain to be able to move forward even if a corrupt committee

Reply:
Otherwise, this looks good.

We might consider checking consistency of all such FullAttestations in shard fork choice. Or maybe even making inconsistency here slashable

Reply:
Can the attestations that voted for the winning shard transition using different `shard_head_root`?? My intuition is no, and that marks the block as invalid.  Say a block consists the following attestations:
```
{slot:100, committe_index:1, shard_transition_root: 0xa, shard_head_root: 0xb}
{slot:100, committe_index:1, shard_transition_root: 0xa, shard_head_root: 0xb}
{slot:100, committe_index:1, shard_transition_root: 0xa, shard_head_root: 0xc}
{slot:100, committe_index:1, shard_transition_root: 0xd, shard_head_root: 0xe}
```
`0xa` gets the crosslink because of 3/4, but the block is bad because the third attestation has `0xc`. Or do we simply not count 0xc, and the vote is 2/4?

Reply:
@djrtwo 
> this could cause the chain to stall if a corrupt committee voted on `ShardTransition()`

I think it has been banned by `verify_empty_shard_transition`?

> We want to add add in `validate_attestation` under `if is_on_time_attestation`: `assert attestation.data.shard_transition_root != hash_tree_root(ShardTransition())`.

Yep, it looks better. 👍 

> We might consider checking consistency of all such FullAttestations in shard fork choice. Or maybe even making inconsistency here slashable

Sounds correct to me.

Btw, to double-check, in [phase 1 v-guide: Broadcast aggregate](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase1/validator.md#broadcast-aggregate), is the `AggregateAndProof.aggregate` type `Attestation` or `FullAttestation`? I suppose the beacon block proposer needs the shard transitions to maintain the [`full_transitions` map](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase1/validator.md#shard-transitions)?

---

@terencechain

Good question!
My intuition is no too, and I think that is a bad block. Moreover, I wonder if we can validate the `shard_head_root` of the 4th attestation as well. But to do it, we need the beacon block to contain all the corresponding `shard_transition` of all attestations... which may blow up the beacon block size...

((keep thinking about it))

Reply:
> I think it has been banned by verify_empty_shard_transition?

A proposer adding `ShardTransition()` for a specific shard is them signalling that the shard in question did not get the requisite support to for a transition. This would not prevent attesters from actually voting on empty shard transition (which could cause weird inconsistencies which we just banned with the new condition)

> Btw, to double-check, in phase 1 v-guide: Broadcast aggregate, is the AggregateAndProof.aggregate type Attestation or FullAttestation?

It would need to be `FullAttestation`.  But I think we might split up attestation data from shard transitions via their own gossipsub topics to prevent duplication so we'll address it there

Reply:
> . But to do it, we need the beacon block to contain all the corresponding shard_transition of all attestations... which may blow up the beacon block size...

yeah, we could but are currently avoiding extra shard transitions. (1) we can ignore incoherent messages in the fork choice, and (2) we could even make this incoherency slashable but not sure it's worth it